vmm/src/device_manager.rs

```diff
@@ -1655,12 +1662,38 @@ impl DeviceManager {
                 )
                 .map_err(DeviceManagerError::Disk)?;
 
-            let mut raw_img = qcow::RawFile::new(file.try_clone().unwrap(), disk_cfg.direct);
+            let image_type =
+                detect_image_type(&mut file).map_err(DeviceManagerError::DetectImageType)?;
 
-            let image_type = qcow::detect_image_type(&mut raw_img)
-                .map_err(DeviceManagerError::DetectImageType)?;
+            // The O_DIRECT flag has some alignment requirements which are not
+            // met by the detect_image_type() function. That's why the flag is
+            // set after the file has been opened and after we found out about
+            // the image type.
+            if disk_cfg.direct {
+                let ret = unsafe { libc::fcntl(file.as_raw_fd(), libc::F_SETFL, libc::O_DIRECT) };
```

Comment:
You lose any existing bits set on the file.

Reply:
```rust
let flags = unsafe { libc::fcntl(file.as_raw_fd(), libc::F_GETFL};
if flags < 0 {
...
}
let ret =  unsafe { libc::fcntl(file.as_raw_fd(),  libc::F_SETFL, flags | libc::O_DIRECT) };
```

Reply:
I got rid of `fnctl` and chose to align the buffers instead.